contracts/DMMPoolV2.sol

```diff
@@ -0,0 +1,378 @@
+            }
+        }
+        uint256 amount0In = newData.reserve0 > data.reserve0 - amount0Out
+            ? newData.reserve0 - (data.reserve0 - amount0Out)
+            : 0;
+        uint256 amount1In = newData.reserve1 > data.reserve1 - amount1Out
+            ? newData.reserve1 - (data.reserve1 - amount1Out)
+            : 0;
+        require(amount0In > 0 || amount1In > 0, "DMM: INSUFFICIENT_INPUT_AMOUNT");
+        uint256 feeInPrecision = verifyBalance(
+            amount0In,
+            amount1In,
+            isAmpPool ? data.vReserve0 : data.reserve0,
+            isAmpPool ? data.vReserve1 : data.reserve1,
+            isAmpPool ? newData.vReserve0 : newData.reserve0,
+            isAmpPool ? newData.vReserve1 : newData.reserve1
+        );
+
+        _update(isAmpPool, newData);
+        emit Swap(msg.sender, amount0In, amount1In, amount0Out, amount1Out, to, feeInPrecision);
+    }
+
+    /// @dev force balances to match reserves
+    function skim(address to) external nonReentrant {
+        token0.safeTransfer(to, token0.balanceOf(address(this)).sub(reserve0));
+        token1.safeTransfer(to, token1.balanceOf(address(this)).sub(reserve1));
+    }
+
+    /// @dev force reserves to match balances
+    function sync() external override nonReentrant {
+        (bool isAmpPool, ReserveData memory data) = getReservesData();
+        bool feeOn = _mintFee(isAmpPool, data);
+        ReserveData memory newData;
+        newData.reserve0 = IERC20(token0).balanceOf(address(this));
+        newData.reserve1 = IERC20(token1).balanceOf(address(this));
+        // update virtual reserves if this is amp pool
+        if (isAmpPool) {
+            uint256 _totalSupply = totalSupply();
+            uint256 b = Math.min(
+                newData.reserve0.mul(_totalSupply) / data.reserve0,
+                newData.reserve1.mul(_totalSupply) / data.reserve1
+            );
+            newData.vReserve0 = Math.max(data.vReserve0.mul(b) / _totalSupply, newData.reserve0);
+            newData.vReserve1 = Math.max(data.vReserve1.mul(b) / _totalSupply, newData.reserve1);
+        }
+        _update(isAmpPool, newData);
+        if (feeOn) kLast = getK(isAmpPool, newData);
+    }
+
+    /// @dev returns data to calculate amountIn, amountOut
+    function getTradeInfo()
+        external
+        virtual
+        override
+        view
+        returns (
+            uint112 _reserve0,
+            uint112 _reserve1,
+            uint112 _vReserve0,
+            uint112 _vReserve1,
+            uint256 feeInPrecision
+        )
+    {
+        // gas saving to read reserve data
+        _reserve0 = reserve0;
+        _reserve1 = reserve1;
+        uint32 _ampBps = ampBps;
+        _vReserve0 = vReserve0;
+        _vReserve1 = vReserve1;
+        if (_ampBps == BPS) {
+            _vReserve0 = _reserve0;
+            _vReserve1 = _reserve1;
+        }
+        feeInPrecision = getFinalFee(FeeFomula.getFee(factorInPrecision), _ampBps);
```

Comment:
why do we use FeeFomula.getFee(factorInPrecision) for feeInPrecision?
It should be fixed when Factory initializes the pool contract

Reply:
Fixed. feeInPrecision is now fixed.